### PR TITLE
[hadoop]: change credential fetcher response type to list instead of single credential

### PR DESCRIPTION
**Pull request by foss-contributor**
## 🥞 Stacked PR
Use this [link](https://github.com/unitycatalog/unitycatalog/pull/1701/files) to review incremental changes.
- [**stack/select-vended-credential-by-location**](https://github.com/unitycatalog/unitycatalog/pull/1701) [[Files changed](https://github.com/unitycatalog/unitycatalog/pull/1701/files)] ← _this PR_
  - [stack/scoped-credential-namespaces](https://github.com/unitycatalog/unitycatalog/pull/1704) [[Files changed](https://github.com/unitycatalog/unitycatalog/pull/1704/files/cc071433385b61e553486e11715f1c15785b65ae..f8de7d7e9a2b33da67ec6fd74e1478a02cae1468)]
    - [stack/encode-scoped-vended-credentials](https://github.com/unitycatalog/unitycatalog/pull/1705) [[Files changed](https://github.com/unitycatalog/unitycatalog/pull/1705/files/f8de7d7e9a2b33da67ec6fd74e1478a02cae1468..a048e325dd999a74cef5e8cfeffa9d8a74bbe31e)]

---------
## Summary

- Change credential fetchers from returning one `GenericCredential` to returning `List<GenericCredential>`. Standard UC responses remain singleton lists, while Delta table and staging-table fetchers preserve all vended credentials.
- Select the longest location-covering credential when multiple credentials are returned, and reject empty responses, missing selection locations, or responses without a covering credential.
- Generalize credential caches for list-valued initial credentials and selected provider credentials. Renewal refetches the list and repeats location selection.
- Cache the selected provider credential using the location-aware delegate filesystem identity while driver-side property construction continues requiring exactly one credential.

## Testing

- Added provider coverage for longest-prefix selection, shared caching, cache bypass, renewal/reselection, and invalid responses.
- Added selection coverage for path boundaries and cloud-scheme aliases.
- Updated UC and Delta fetcher tests for ordered list-valued responses and updated dependent cache, property, and Spark fixtures.

## Issue

https://github.com/unitycatalog/unitycatalog/issues/1694

### CODE REVIEW

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java [hunk 1]
```diff
@@ -188,9 +192,14 @@ private static Map<String, String> createCredProps(
       // Unsupported scheme: skip the credential fetch entirely and return no props.
       return Collections.emptyMap();
     }
-    GenericCredential cred =
-        fetchGenericCredential(
+    List<GenericCredential> credentials =
+        fetchGenericCredentials(
             hadoopConf, apiClient, catalogUri, tokenProvider, appVersions, credId);
+    Preconditions.checkState(
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java:199]: Rejecting `size != 1` breaks today’s Delta behavior. Fetchers used to `selectForLocation(credId.location(), …)` before returning one credential; now a multi-cred response fails planning. Issue #1694 says multi-cred responses already happen and the connector currently picks the covering one.

Why fail here instead of selecting with the Delta/staging `location` already on `CredId`, until the encode PR lands? If this layer must not ship alone, please say that explicitly in the PR.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java [hunk 2]
```diff
@@ -82,10 +85,25 @@ private GenericCredential renewCredential() throws ApiException {
...
+      return fetchAndSelectCredential();
     }
   }
+
+  private GenericCredential fetchAndSelectCredential() throws ApiException {
+    List<GenericCredential> credentials = genericCredentialFetcher().createCredentials();
+    Preconditions.checkState(!credentials.isEmpty(), "No vended credential was returned.");
+    if (credentials.size() == 1) {
+      return credentials.get(0);
+    }
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:108]: Singleton responses skip coverage checks (`singleCredentialIsUsedWithoutMatchingLocation`). Old Delta path failed when the only credential did not cover the location. Why relax that for Delta too?

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java [hunk 3]
```diff
@@ -82,10 +85,25 @@ private GenericCredential renewCredential() throws ApiException {
...
+    Preconditions.checkState(
+        cacheKey.prefix() != null,
+        "Multiple credentials were vended but no location is set to select one.");
+    return CredentialUtil.selectForLocation(cacheKey.prefix(), credentials);
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:115]: Renewal selects via `FileSystemCredId.prefix()` (`UC_CREDENTIAL_PREFIX_KEY`), not `DeltaTableCredId.location()` / `UC_DELTA_LOCATION_KEY`. What breaks if multi-cred renewal runs with Delta location set but prefix unset?

**Review by openinx — COMMENTED** [review]
Left comments on the incremental stacked range (`8a4c0ff1..436d76d5`).

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredPropsUtilTest.java [hunk 4]
```diff
@@ -27,6 +31,31 @@ void credContextIdVariesByCatalogUriSchemeAndAuth() {
         .isNotEqualTo(base);
   }
 
+  @Test
+  void multipleVendedCredentialsAreRejectedOnDriver() {
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredPropsUtilTest.java]: nit: rename this test so it does not say "driver".

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredentialUtil.java [hunk 5]
```diff
@@ -56,34 +56,31 @@ public static GenericCredential toGenericCredential(DeltaStorageCredential cred)
...
     int bestLen = -1;
-    for (DeltaStorageCredential c : creds) {
-      if (c == null || c.getPrefix() == null || !prefixCovers(location, c.getPrefix())) {
+    for (GenericCredential cred : creds) {
+      String prefix = cred.prefix();
+      if (prefix == null || !prefixCovers(location, prefix)) {
         continue;
       }
-      int len = stripTrailingSlashes(c.getPrefix()).length();
+      int len = stripTrailingSlashes(prefix).length();
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredentialUtil.java:77]: The prefixCovers already stripped the trailing slash for the `prefix`, and here we do another trailing slash again, which seems unnecessary.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/CredentialCache.java [hunk 6]
```diff
@@ -74,41 +78,43 @@ int size() {
...
-   * later readers observe the same renewal behavior.
-   */
-  public static class RenewableCredential {
+  /** Decides whether {@code value} is close enough to expiry that it should be renewed. */
+  @FunctionalInterface
+  public interface Staleness<T> {
+    boolean readyToRenew(T value, Clock clock, long renewalLeadTimeMillis);
+  }
+
+  public static class RenewableCredential<T> {
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/CredentialCache.java]: Do you think making this class as `abstracted` and provide the `readyToRenew()` will actually simply the interface that we want to expose to the caller ?  the current PR i see two problems here: 

1.  we are passing the dup `Clock` and `renewalLeadTimeMillis` for both `Staleness` interface and `RenewalableCredential` class, and what if those are not matching ?? then it's a risk to be buggy. 

2.  What if the `Stateless<T>` does not match the `RenewableCredential` constructor ?  that's also a risk.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialFetcher.java [hunk 7]
```diff
@@ -24,7 +25,11 @@
  * <p><b>Internal API — not for external use. May change without notice.</b>
  */
 public interface GenericCredentialFetcher {
-  GenericCredential createCredential() throws ApiException;
+  /**
+   * Fetches the vended credentials for this fetcher's scope. A single-credential API returns a
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialFetcher.java]: > A single-credential API returns a
   * one-element list.

This is unncessary and seems to be verbose.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java [hunk 8]
```diff
@@ -81,10 +85,25 @@ private GenericCredential renewCredential() throws ApiException {
...
-      return genericCredentialFetcher().createCredential();
+      return fetchAndSelectCredential();
     }
   }
+
+  private GenericCredential fetchAndSelectCredential() throws ApiException {
+    List<GenericCredential> credentials = genericCredentialFetcher().createCredentials();
+    Preconditions.checkState(!credentials.isEmpty(), "No vended credential was returned.");
+    if (credentials.size() == 1) {
+      return credentials.get(0);
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:107]: the intellij actually suggest to use `credentials.getFirst()`.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcher.java [hunk 9]
```diff
@@ -39,7 +41,8 @@ public GenericCredential createCredential() throws ApiException {
         id.catalog(),
         id.schema(),
         id.table());
-    return CredentialUtil.toGenericCredential(
-        CredentialUtil.selectForLocation(credId.location(), response.getStorageCredentials()));
+    return response.getStorageCredentials().stream()
+        .map(CredentialUtil::toGenericCredential)
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcher.java:53]: few comments for the CredentialUtil.toGenericCredential, but unrelated to this PR:

1. `toGenericCredential`, there is no `isGcsCred` check when we assume it's a `GcsCredential`.   and for those unexpected cases, we should throw. 

2.  `requireSingleCloudConfig`,  we'd better to reuse all the existing `isS3Config -> isS3Cred` / `isAzureConfig -> isAzureCred`/ `isGcsCred` methods. 

3. `toGenericCredential`,  i suggest to define those `isXXCred` for `TemporaryCredentials` also.  that help us to use the same methods as the `DeltaStorageCredential`.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java [hunk 10]
```diff
@@ -6,14 +6,15 @@
...
+import java.util.List;
 import java.util.UUID;
+import java.util.stream.Collectors;
 
 /** Adapts the UC Delta staging table credentials SDK API for Hadoop token providers. */
 final class UCDeltaStagingTableCredentialFetcher implements GenericCredentialFetcher {
 
   private final DeltaTemporaryCredentialsApi api;
   private final UUID stagingTableId;
-  private final String stagingTableLocation;
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java:16]: If you are removing the `stagingTableLocation` here, why don't we also remove the `location` in the `DeltaStagingTableCredId` and even the `location` parameter in `createDeltaStagingTableCredProps` ?  

and that will even bring us a new issue, do we still need to support the public API that `unitycatalog-hadoop` module provides since `location` will be never used  ?  then we will need to deprecate the existing one, and add another one with only the `stagingTableId` parameter.

```
/**
     * Builds Hadoop properties for a <em>Delta staging table</em> using the UC Delta staging table
     * credentials API ({@code GET /delta/v1/staging-tables/{table_id}/credentials}).
     *
     * @param stagingTableId the staging table UUID
     * @param location the staging table storage location, used to select the matching storage
     *     credential from the response
     * @return unmodifiable map; empty if the scheme is unrecognized
     * @throws IllegalArgumentException if {@code stagingTableId}, {@code location}, or {@code
     *     tokenProvider} is missing
     * @throws ApiException if the credential fetch from the UC Delta API fails
     */
    public Map<String, String> buildForStagingTable(String stagingTableId, String location)
        throws ApiException {
```

consider all the context,  I suggest you open another separate PR to finish this removal.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java (hunk 1 above)
Comment (openinx) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java:199]: If you do want to make the 1 cred support in this PR, pls explicitly add the TODO in those line.

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/AbfsVendedTokenProviderTest.java [hunk 11]
```diff
@@ -20,12 +20,22 @@ protected AbfsVendedTokenProvider createTestProvider(
...
+  @Override
+  protected AbfsVendedTokenProvider createTestProvider(
+      Configuration conf, GenericCredentialFetcher fetcher) {
+    return new TestAbfsVendedTokenProvider(conf, fetcher);
+  }
+
   static class TestAbfsVendedTokenProvider extends AbfsVendedTokenProvider {
     private final GenericCredentialFetcher credentialFetcher;
 
     TestAbfsVendedTokenProvider(Configuration conf, TemporaryCredentialsApi mockApi) {
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/AbfsVendedTokenProviderTest.java]: why don't you remove this constructor directly, and making all the mock method reuse the below override constructor directly ?  as the `TestAbfsVendedTokenProvider(Configuration conf, GenericCredentialFetcher credentialFetcher)` is always more abstracted to implement any kind of mocking credential ?

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredPropsUtilTest.java (hunk 4 above)
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredPropsUtilTest.java]: This test sets `CredPropsUtil.genericCredFetcherFactory` and never resets it. `CredPropsBaseTest` / `CredPropsCacheTest` / `CredPropsRoundTripTest` all restore the factory in `@AfterEach`. Please add the same reset, or later tests in the same JVM can get poisoned.

nit: also rename so the method does not say "driver".

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredentialUtilTest.java [hunk 12]
```diff
@@ -150,20 +110,27 @@ void selectorNormalizesTrailingSlashes() {
...
+    List<GenericCredential> creds =
+        Arrays.asList(credAt(null), credAt("s3://other"), credAt("s3://bucket/sibling"));
     assertThatThrownBy(() -> CredentialUtil.selectForLocation("s3://bucket/t", creds))
         .isInstanceOf(IllegalArgumentException.class)
-        .hasMessageContaining("No UC Delta credential matched");
+        .hasMessageContaining("No vended credential covers location");
+    assertThatThrownBy(() -> CredentialUtil.selectForLocation("S3A://bucket/t", creds))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("No vended credential covers location");
   }
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredentialUtilTest.java]: We dropped several `selectForLocation` edge-case tests that still matter for the new `List<GenericCredential>` API:
- empty list (and null list if that can reach it)
- single credential that does not cover the location (or cover this only via the provider shortcut, intentionally)
- null / prefix-less entries in a multi-cred list (old code skipped nulls; new code NPEs on null `cred`)

Please restore the ones that still match intended behavior, or add provider/fetcher tests that replace them.

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredentialUtilTest.java [hunk 13]
```diff
@@ -150,20 +110,27 @@ void selectorNormalizesTrailingSlashes() {
...
-    List<DeltaStorageCredential> creds =
-        Arrays.asList(credAt("s3://other"), credAt("s3://bucket/sibling"));
+  void selectForLocationThrowsWhenNoPrefixCovers() {
+    List<GenericCredential> creds =
+        Arrays.asList(credAt(null), credAt("s3://other"), credAt("s3://bucket/sibling"));
     assertThatThrownBy(() -> CredentialUtil.selectForLocation("s3://bucket/t", creds))
         .isInstanceOf(IllegalArgumentException.class)
-        .hasMessageContaining("No UC Delta credential matched");
+        .hasMessageContaining("No vended credential covers location");
+    assertThatThrownBy(() -> CredentialUtil.selectForLocation("S3A://bucket/t", creds))
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/CredentialUtilTest.java]: The `S3A://…` case does not prove scheme non-aliasing: this fixture has no covering `s3://bucket/t` credential, so failure is expected even with scheme normalization. Did you mean something like one cred at `s3://bucket/t` and location `s3a://bucket/t/file`, expecting no match?

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/BaseTokenProviderTest.java [hunk 14]
```diff
@@ -70,6 +84,119 @@ public void after() {
...
+  }
+
+  @Test
+  public void singleCredentialIsUsedWithoutMatchingLocation() {
+    GenericCredential only = newGenericCred("only", Long.MAX_VALUE, location("/other"));
+
+    T provider = provider(location("/t/file"), only);
+
+    assertThat(provider.accessCredentials()).isSameAs(only);
+  }
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/BaseTokenProviderTest.java:147]: This locks in “singleton skips coverage.” Is that the intended contract for Delta too, or only UC REST? If intentional, fine; if not, the test is encoding the wrong behavior.

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcherTest.java [hunk 15]
```diff
@@ -44,17 +45,70 @@ void createCredentialCallsDeltaApiWithCredIdFieldsAndReturnsCredential() throws
...
     assertThat(cred.secretAccessKey()).isEqualTo("sk");
     assertThat(cred.sessionToken()).isEqualTo("st");
     assertThat(cred.expirationTimeMillis()).isEqualTo(789L);
+    assertThat(cred.prefix()).isEqualTo("s3://bucket/events");
     verify(api)
         .getTableCredentials(DeltaCredentialOperation.READ_WRITE, "main", "default", "events");
   }
 
+  @Test
+  void createCredentialsReturnsAllVendedCredentialsInOrder() throws Exception {
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcherTest.java:66]: nit: table vs staging `createCredentialsReturnsAllVendedCredentialsInOrder` tests are near-duplicates; a small shared helper would shrink noise.

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcherTest.java [hunk 16]
```diff
@@ -44,17 +45,70 @@ void createCredentialCallsDeltaApiWithCredIdFieldsAndReturnsCredential() throws
...
+    AwsCredential cred1 = (AwsCredential) creds.get(0);
+    assertThat(cred1.accessKeyId()).isEqualTo("ak1");
+    assertThat(cred1.prefix()).isEqualTo("s3://bucket/events");
+    AwsCredential cred2 = (AwsCredential) creds.get(1);
+    assertThat(cred2.accessKeyId()).isEqualTo("ak2");
+    assertThat(cred2.prefix()).isEqualTo("s3://bucket/events/child");
+  }
+
   @Test
   void createCredentialRejectsMissingDeltaCredentialsResponse() throws Exception {
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcherTest.java]: nit: method still named `createCredentialRejects…` after the API rename to `createCredentials`.

Also missing negative coverage for null/empty `getStorageCredentials()` (pairs with the production null-check gap).

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcherTest.java [hunk 17]
```diff
@@ -30,17 +31,64 @@ void createCredentialCallsDeltaStagingApiAndReturnsCredential() throws Exception
...
+    AwsCredential cred1 = (AwsCredential) creds.get(0);
+    assertThat(cred1.accessKeyId()).isEqualTo("ak1");
+    assertThat(cred1.prefix()).isEqualTo(LOCATION);
+    AwsCredential cred2 = (AwsCredential) creds.get(1);
+    assertThat(cred2.accessKeyId()).isEqualTo("ak2");
+    assertThat(cred2.prefix()).isEqualTo(LOCATION + "/child");
+  }
+
   @Test
   void createCredentialRejectsNullResponse() throws Exception {
```
Comment (openinx) [inline connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcherTest.java]: nit: rename `createCredentialRejectsNullResponse` to match `createCredentials`.

Same gap as the table fetcher: please add coverage for null/empty `getStorageCredentials()`.

**Review by openinx — COMMENTED** [review]
Test-focused follow-up comments.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java (hunk 3 above)
Comment (openinx) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:115]: Pls explicitly add the comment to attach the context that we've discussed offline. I mean for multip cred cases, we should set the `fs.unitycatalog.credential.prefix` explicitly in the `CredPropsBuilder.build()`.  while for 1 single cred, we also should set it  but we want to skip the prefix selection directly and use the only one cred.

Comment (openinx) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:115]: if possible,  pls also context in the `UC_CREDENTIAL_PREFIX_KEY`

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcher.java (hunk 9 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaGenericCredentialFetcher.java:53]: I'll file this as follow up PR

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java (hunk 10 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java:16]: Let's do a follow up PR for future release.

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/BaseTokenProviderTest.java (hunk 14 above)
Comment (foss-contributor) [reply connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/BaseTokenProviderTest.java:147]: Yes, locking in contract that when there is only credential, we use it regardless of if it actually covers the requested storage location. I'll update the GitHub Issue to reflect this.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java (hunk 2 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:108]: This is the intended new behavior, updated it in the GitHub issue.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java (hunk 3 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java:115]: Requirement is if multi cred then must set, otherwise optional. Updated comments. I'll update the GitHub issue as well.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java (hunk 1 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java:199]: Let's keep each PR self contained; Will fix

File: connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/AbfsVendedTokenProviderTest.java (hunk 11 above)
Comment (foss-contributor) [reply connectors/hadoop/src/test/java/io/unitycatalog/hadoop/internal/auth/AbfsVendedTokenProviderTest.java]: Done, cleaned up for GCP and AWS as well

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/CredentialCache.java (hunk 6 above)
Comment (openinx) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/CredentialCache.java]: Hi @foss-contributor ,  can you take a look at this comment also ?

Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/CredentialCache.java]: Fixing it right now

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java [hunk 18]
```diff
@@ -80,11 +84,34 @@ private GenericCredential renewCredential() throws ApiException {
     if (credCacheEnabled) {
       return globalCache.access(
           cacheKey,
-          () ->
-              new RenewableCredential(
-                  renewalLeadTimeMillis, clock, genericCredentialFetcher().createCredential()));
+          () -> {
+            GenericCredential credential = fetchAndSelectCredential();
+            return new RenewableCredential<GenericCredential>(credential) {
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/GenericCredentialProvider.java]: nit:  this `GenericCredential` can be removed in the class new, since it will automatically infer the class type.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java [hunk 19]
```diff
@@ -242,14 +251,21 @@ private static GenericCredential fetchGenericCredential(
...
-        () ->
-            new RenewableCredential(
-                renewalLeadTimeMillis,
-                ClockUtil.resolveClock(hadoopConf),
-                createCredential(apiClient, catalogUri, tokenProvider, appVersions, credId)));
+        () -> {
+          Clock clock = ClockUtil.resolveClock(hadoopConf);
+          List<GenericCredential> credentials =
+              createCredentials(apiClient, catalogUri, tokenProvider, appVersions, credId);
+          return new RenewableCredential<List<GenericCredential>>(credentials) {
```
Comment (openinx) [inline connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/CredPropsUtil.java]: nit: This `List<GenericCredential>` can also be removed now.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java (hunk 10 above)
Comment (openinx) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java:16]: then for this PR, we will need to revert the `stagingTableLocation` deletion,  pls don't do the unnecessary removal in this PR, and let's defer to the follow-up PR.

**Review by openinx — APPROVED** [review]
Looks good to me now, but pls address the new 3 minor comments.  I will merge it once everything is fine.   Thanks @foss-contributor for the contribution.

File: connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java (hunk 10 above)
Comment (foss-contributor) [reply connectors/hadoop/src/main/java/io/unitycatalog/hadoop/internal/auth/UCDeltaStagingTableCredentialFetcher.java:16]: yes
